### PR TITLE
Add two-phase migration for SSH transport method

### PR DESCRIPTION
**Pull request by fdupont-redhat**
This pull request adds two-phase migration for SSH transport method.
It also sets the default user to `root` as we connect to ESXi hosts, not vCenter.

RHBZ: https://bugzilla.redhat.com/show_bug.cgi?id=1817096

### CODE REVIEW

File: wrapper/pre_copy.py [hunk 1]
```diff
@@ -202,7 +204,13 @@ def auth_cb(cred, _):
 
         cred_info = [[libvirt.VIR_CRED_AUTHNAME, libvirt.VIR_CRED_PASSPHRASE],
                      auth_cb, None]
-        conn = libvirt.openAuth(self._uri, cred_info)
+
+        if re.compile('^ssh://').match(self._uri):
+            uri = "esx://%s@%s/?no_verify=1" % (self.user, self.server)
```
Comment (nyoxi) [inline wrapper/pre_copy.py]: So far we have been using `insecure_connection` to disable SSL verification only for output connections, but it may make sense to extend this also to input connection here.

Alternatively we can keep requesting `no_verify` in input URI. You can then check if it is present in the query part of the input URI and propagate it here.

File: wrapper/pre_copy.py [hunk 2]
```diff
@@ -716,7 +763,8 @@ def _start_nbdkits(self, final):
                 filters.remove(filt)
 
         for disk in self.disks:
-            cmd = self._get_nbdkit_cmd(disk, filters, final)
+            cmd = getattr(self, "_get_nbdkit_cmd_%s" % self._transport_method)(
+                    disk, filters, final)
```
Comment (nyoxi) [inline wrapper/pre_copy.py]: let's say `self._transport_method` contains something invalid where will the `AttributeError` generated by `getattr()` be handled?

File: wrapper/pre_copy.py [hunk 3]
```diff
@@ -114,7 +114,7 @@ def __init__(self, data):
 
         self.server = uri.hostname
         self.port = uri.port
-        self.user = 'administrator@vsphere.local'
+        self.user = 'root'
```
[the remark this answers is not in the document]
Comment (fdupont-redhat) [reply wrapper/pre_copy.py]: We don't use `vpx` anymore.

File: wrapper/pre_copy.py [hunk 4]
```diff
@@ -137,6 +137,8 @@ def __init__(self, data):
                 self.insecure = False
             else:
                 raise ValueError('Invalid value for "no_verify"')
+        elif data['transport_method'] == 'ssh':
+            self.insecure = True
```
[the remark this answers is not in the document]
Comment (fdupont-redhat) [reply wrapper/pre_copy.py]: Because with SSH URI, there's no `no_verify` parameter. So, we never go into the condition above.
And I'm forcing insecure connection as default in the case of SSH, because `no_verify` is always true with VDDK when called by CloudForms.

File: wrapper/pre_copy.py (hunk 1 above)
Comment (fdupont-redhat) [reply wrapper/pre_copy.py]: The input URI for SSH doesn't have `no_verify` parameter.

File: wrapper/pre_copy.py (hunk 2 above)
Comment (fdupont-redhat) [reply wrapper/pre_copy.py]: `transport_method` value is checked here: https://github.com/ManageIQ/manageiq-v2v-conversion_host/blob/master/wrapper/virt_v2v_wrapper.py#L392. So at this stage it will already have been caught.

File: wrapper/pre_copy.py (hunk 4 above)
Comment (nertpinx) [inline wrapper/pre_copy.py]: Is the reason for this that connecting with `esx://` won't have any valid certificate? Or is it possible to have a setup where certificate validation would work securely?

File: wrapper/pre_copy.py (hunk 3 above)
Comment (nertpinx) [inline wrapper/pre_copy.py]: I know I said I'm fine with it, but I figured it would make more sense (for users/developers, not management apps) to have this value based on the scheme, `esx` -> `root` and `vpx` -> `administrator@vsphere.local` would make sense.

File: wrapper/pre_copy.py [hunk 5]
```diff
@@ -202,7 +204,13 @@ def auth_cb(cred, _):
 
         cred_info = [[libvirt.VIR_CRED_AUTHNAME, libvirt.VIR_CRED_PASSPHRASE],
                      auth_cb, None]
-        conn = libvirt.openAuth(self._uri, cred_info)
+
+        if re.compile('^ssh://').match(self._uri):
```
Comment (nertpinx) [inline wrapper/pre_copy.py]: All examples I've seen were using `esx://user@host/`, is the `ssh://` uri used as well somewhere?

File: wrapper/pre_copy.py [hunk 6]
```diff
@@ -340,6 +349,12 @@ def __init__(self, nbd, disk, tmp_dir):
         self.proc_qemu = None
         self.overlay = None
 
+        datastore = disk.backing.datastore
+        datastore_mountpoint = datastore.summary.url.replace("ds://", "")
+        self.absolute_path = "%s/%s" % (
+                datastore_mountpoint,
+                disk.backing.fileName.replace("[%s] " % datastore.name, ""))
```
Comment (nertpinx) [inline wrapper/pre_copy.py]: Something like:
```
fileName = disk.backing.fileName
ds_id = "[%s] " % datastore.name
if fileName.startswith(ds_id):
    fileName = fileName[len(ds_id):]
```
or even, if you are sure both strings will always start with the proper prefix, something just this:
```
datastore.summary.url[len("ds://"):] + '/' + disk.backing.fileName[len('[%s] ' % datastore.name):]
```
would be safer, just in case there is some mr. smartypants who has `[datastore] ` or as part of the filename.

File: wrapper/pre_copy.py [hunk 7]
```diff
@@ -660,7 +677,7 @@ def get_xml(self):
             f.write(self._fix_disks(self.vmware.get_domxml()))
         return xmlfile
 
-    def _get_nbdkit_cmd(self, disk, filters, final):
+    def _get_nbdkit_cmd_vmdk(self, disk, filters, final):
```
Comment (nertpinx) [inline wrapper/pre_copy.py]: did you mean `_get_nbdkit_cmd_vddk`?

File: wrapper/pre_copy.py [hunk 8]
```diff
@@ -698,6 +715,36 @@ def _get_nbdkit_cmd(self, disk, filters, final):
...
+            '-U', disk.sock,
+            '-P', disk.pidfile,
+            '--exit-with-parent',
+            '--readonly',
+            '--foreground',
+            '--exportname=/',
+        ] + [
+            '--filter=' + f for f in filters
+        ] + [
+            '--filter=log',
```
Comment (nertpinx) [inline wrapper/pre_copy.py]: Most of this (and some stuff below) should be similar in both functions, no need to duplicate the code, just add it to something like `_get_nbdkit_cmd_common()` or something and call it from both.

File: wrapper/pre_copy.py [hunk 9]
```diff
@@ -698,6 +715,36 @@ def _get_nbdkit_cmd(self, disk, filters, final):
...
+            'ssh',
+            # pylint: disable=protected-access
```
Comment (nertpinx) [inline wrapper/pre_copy.py]: Accidental leftover?

File: wrapper/pre_copy.py [hunk 10]
```diff
@@ -698,6 +715,36 @@ def _get_nbdkit_cmd(self, disk, filters, final):
...
+            'host=%s' % self.vmware.server,
+            'password=+%s' % self._vmware_password_file,
+            'path=%s' % disk.absolute_path.replace('.vmdk', '-flat.vmdk'),
```
Comment (nertpinx) [inline wrapper/pre_copy.py]: Something like this would be safer:
```suggestion
            'path=%s' % '-flat'.join(os.path.splitext(disk.absolute_path.replace()),
```

Comment (nertpinx) [reply wrapper/pre_copy.py]: Or, if there might be other files than just `.vmdk`:
```
root, ext = os.path.splitext(disk.absolute_path)
if ext == 'vmdk':
    root = root + '-flat'
path = root + '.' + ext
```

Comment (nertpinx) [reply wrapper/pre_copy.py]: You can also doit right away when assigning into the absolute path, so that here it is really just constructing a command line, not doing other logic.

File: wrapper/pre_copy.py (hunk 2 above)
Comment (nertpinx) [inline wrapper/pre_copy.py]: Passing the transport method as a parameter should be enough and slightly more readable, I think. Actually, you already have it saved in the object already.

File: wrapper/tests/test_pre_copy.py [hunk 11]
```diff
@@ -7,11 +7,12 @@ class TestPreCopy(unittest.TestCase):
 
     basic_data = {
         'output_format': 'raw',
-        'vmware_uri': 'vpx://example.com',
+        'vmware_uri': 'esx://esx1.example.com',
```
Comment (nertpinx) [inline wrapper/tests/test_pre_copy.py]: You are removing code coverage by testing only the esx transport even though the code supports more than that.

File: wrapper/pre_copy.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (nertpinx) [reply wrapper/pre_copy.py]: Well you in particular do not, but why removing the support for testing it easily, at least for me, for example. Especially when it when it works perfectly?

File: wrapper/pre_copy.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (nertpinx) [reply wrapper/pre_copy.py]: I understand that the setup most seen in the world is isolated network with self-signed certificates. But defaulting to not even trying to check them, just in case someone as a proper secure setup seems a bit backwards to me.

**Comment by nertpinx** [comment]
Oh, I had it opened for a while and missed that there is already one review, sorry for that.

File: wrapper/pre_copy.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (nyoxi) [reply wrapper/pre_copy.py]: The fact that `vpx` is not used by ManageIQ does not mean it cannot be used from command line.

File: wrapper/pre_copy.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (nyoxi) [reply wrapper/pre_copy.py]: Again, the fact that MIQ/CFME enforces `no_verify` everywhere does not mean it's the only way (especially when in this case it is really bad way).

File: wrapper/pre_copy.py (hunk 1 above)
Comment (nyoxi) [reply wrapper/pre_copy.py]: I don't think there's any specification around that. Unless it breaks virt-v2v (in case it accidentally sees it) I don't see a reason why we couldn't add that at our end.

File: wrapper/pre_copy.py (hunk 2 above)
Comment (nyoxi) [reply wrapper/pre_copy.py]: I am not saying this needs fixing (just yet), but the check is in a separate part of code. Assuming we change the check or behavior later to support more protocols and we forget to add a new method here, where will the error be caught?

File: wrapper/pre_copy.py (hunk 5 above)
Comment (nyoxi) [reply wrapper/pre_copy.py]: `ssh://` is virt-v2v construct

File: wrapper/pre_copy.py (hunk 9 above)
Comment (fdupont-redhat) [reply wrapper/pre_copy.py]: It was already there. And if I remove it, tox tells me ```Access to a protected member _moId of a client class (protected-access)```, because we access the `_moId` attribute of the VM.

File: wrapper/pre_copy.py (hunk 10 above)
Comment (fdupont-redhat) [reply wrapper/pre_copy.py]: That should always be a `.vmdk` file. And the logic here is to derive the flat file name from the absolute path, as the absolute path point to the generic vmdk file.

File: wrapper/pre_copy.py (hunk 2 above)
Comment (fdupont-redhat) [reply wrapper/pre_copy.py]: Validating the transport methods should be done in the `validate_data` method for the provider. Currently, there's only once source provider, VMware, so it's still done in `virt-v2v-wrapper.py`. IMO, the VMware provider would deserve its own `VmwareHost` class. But the `hosts.py` becomes extra large, so we should work to split the code, maybe with more interfaces/factories.

File: wrapper/tests/test_pre_copy.py (hunk 11 above)
Comment (fdupont-redhat) [reply wrapper/tests/test_pre_copy.py]: Ok. Then I've added test for SSH.

**Comment by fdupont-redhat** [comment]
@nyoxi @nertpinx are you waiting for more changes from me?

File: wrapper/pre_copy.py [hunk 12]
```diff
@@ -202,7 +209,15 @@ def auth_cb(cred, _):
 
         cred_info = [[libvirt.VIR_CRED_AUTHNAME, libvirt.VIR_CRED_PASSPHRASE],
                      auth_cb, None]
-        conn = libvirt.openAuth(self._uri, cred_info)
+
+        if self._uri.startswith('ssh://'):
+            uri = "esx://%s@%s/" % (self.user, self.server)
```
Comment (nyoxi) [inline wrapper/pre_copy.py]: You need to `quote()` the user here

File: wrapper/virt_v2v_wrapper.py [hunk 13]
```diff
@@ -63,9 +63,9 @@ def prepare_command(data, v2v_caps, agent_sock=None):
             '--no-copy',
         ])
     else:
-        v2v_args.append(data['vm_name'])
```
Comment (nyoxi) [inline wrapper/virt_v2v_wrapper.py]: For SSH we've always kept the `ssh://` URI in `vm_name`, why this change?

Comment (fdupont-redhat) [reply wrapper/virt_v2v_wrapper.py]: Because I introduce a `vmware_uri` field in https://github.com/ManageIQ/manageiq/pull/20011. It will be more consistent with what we have for VDDK.

Comment (fdupont-redhat) [reply wrapper/virt_v2v_wrapper.py]: I've updated the documentation accordingly.
